server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1023,10 +1081,58 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                         deleteSnapshotsFromRepository(entry, newState.nodes().getMinNodeVersion());
                     }
                 }
+
+                for (SnapshotsInProgress.Entry entry : snapshotsToBwCFail) {
+                    failSnapshotCompletionListeners(
+                            entry.snapshot(),
+                            new ConcurrentSnapshotExecutionException(
+                                    entry.snapshot(),
+                                    "one or more nodes of a version older than the compatibility version of this snapshot joined the " +
+                                            "cluster while it was running"
+                            )
+                    );
+                }
+                for (SnapshotDeletionsInProgress.Entry entry : deletionsToBwCFail) {
+                    failListenersIgnoringException(
+                            snapshotDeletionListeners.remove(entry.uuid()),
+                            new ConcurrentSnapshotExecutionException(
+                                    entry.repository(),
+                                    "_na",
+                                    "one or more nodes of a version incompatible to this repository operation joined the cluster"
+                            )
+                    );
+                    repositoryOperations.finishDeletion(entry.uuid());
+                }
             }
         });
     }
 
+    /**
+     * Ensures that no snapshot operations that are incompatible with the current cluster node versions can be found in the cluster state
+     * at the moment and throws an exception if any are found. These incompatible tasks can arise if a node with a version older than the
+     * previous minimum node version in the cluster joins the cluster after a snapshot-, clone- or delete-operation has been started.
+     * These tasks will be removed from the cluster state by {@link #processExternalChanges(boolean, boolean)} but in rare instances if a
+     * snapshot operation is requested immediately after an old version node joins the cluster this state can be encountered.
+     * This method is called before any new operations are added for a repository to ensure that no tasks are created for the repo until
+     * the existing broken tasks are removed from the cluster state.
+     *
+     * @param repoName repository name
+     * @param state    current cluster state
+     */
+    private static void ensureHealthyTaskVersions(String repoName, ClusterState state) {
```

Comment:
The approach of just checking and throwing via this one in delete, create and clone operations is necessary to guard against all races due to the ordering of operations we work with. We set up the state update that fails and removes incompatible snapshots as a result of applying the CS to the snapshots service but we could theoretically see a task be enqueued before the cleanup task via `processExternalChanges`.

Reply:
SLM doesn't really "retry" in the classical sense. It just does it's jobs on the configured schedule. So if this kills a snapshot/clone then in the next snapshot period it retries, same for its deletes. This isn't different from other exceptions in that regard.